<link rel="stylesheet" href="../../..//default.css">
<script src="../../..//highlight.pack.js"></script> 
<script>hljs.initHighlightingOnLoad();</script>
<html><pre><code class='java'>
<a href="https://github.com/calico/basenji/blob/master/bin/sonnet_satg_bed.py#L52">GitHubLink</a>


<a href="https://github.com/maldil/basenji/blob/master/bin/sonnet_satg_bed.py#L52">GitMyHubLink</a>

&#47&#47!/usr/bin/env python
&#47&#47 Copyright 2017 Calico LLC
&#47&#47
&#47&#47 Licensed under the Apache License, Version 2.0 (the "License");
&#47&#47 you may not use this file except in compliance with the License.
&#47&#47 You may obtain a copy of the License at
&#47&#47
&#47&#47     https://www.apache.org/licenses/LICENSE-2.0
&#47&#47
&#47&#47 Unless required by applicable law or agreed to in writing, software
&#47&#47 distributed under the License is distributed on an "AS IS" BASIS,
&#47&#47 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
&#47&#47 See the License for the specific language governing permissions and
&#47&#47 limitations under the License.
&#47&#47 =========================================================================
from __future__ import print_function

from optparse import OptionParser

import gc
import json
import os
import pdb
import pickle
from queue import Queue
import random
import sys
from threading import Thread

import h5py
import numpy as np
import pandas as pd
import pysam
import tensorflow as tf

if tf.__version__[0] == &quot1&quot:
  tf.compat.v1.enable_eager_execution()

from basenji import bed
from basenji import dna_io
from basenji import seqnn

&quot&quot&quot
sonnet_satg_bed.py

Perform an in silico saturation mutagenesis of sequences in a BED file.
&quot&quot&quot

&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
&#47&#47 main
&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
def main():
  <a id="change">usage = &quotusage: %prog [options] &lt;model&gt; &lt;bed_file&gt;&quot</a>
  <a id="change">parser = OptionParser(usage)</a>
  parser.add_option(&quot-c&quot, dest=&quotslice_center&quot,
      default=False, action=&quotstore_true&quot,
      help=&quotSlice center position(s) for gradient [Default: %default]&quot)
  parser.add_option(&quot-d&quot, dest=&quotmut_down&quot,
      default=0, type=&quotint&quot,
      help=&quotNucleotides downstream of center sequence to mutate [Default: %default]&quot)
  parser.add_option(&quot-f&quot, dest=&quotgenome_fasta&quot,
      default=&quot%s/data/hg38.fa&quot % os.environ[&quotBASENJIDIR&quot],
      help=&quotGenome FASTA for sequences [Default: %default]&quot)
  parser.add_option(&quot-l&quot, dest=&quotmut_len&quot,
      default=0, type=&quotint&quot,
      help=&quotLength of center sequence to mutate [Default: %default]&quot)
  parser.add_option(&quot-o&quot, dest=&quotout_dir&quot,
      default=&quotsat_mut&quot, help=&quotOutput directory [Default: %default]&quot)
  parser.add_option(&quot-p&quot, dest=&quotprocesses&quot,
      default=None, type=&quotint&quot,
      help=&quotNumber of processes, passed by multi script&quot)
  parser.add_option(&quot--rc&quot, dest=&quotrc&quot,
      default=False, action=&quotstore_true&quot,
      help=&quotEnsemble forward and reverse complement predictions [Default: %default]&quot)
  parser.add_option(&quot--shifts&quot, dest=&quotshifts&quot,
      default=&quot0&quot, type=&quotstr&quot,
      help=&quotEnsemble prediction shifts [Default: %default]&quot)
  parser.add_option(&quot--species&quot, dest=&quotspecies&quot,
      default=&quothuman&quot)
  parser.add_option(&quot-t&quot, dest=&quottargets_file&quot,
      default=None, type=&quotstr&quot,
      help=&quotFile specifying target indexes and labels in table format&quot)
  parser.add_option(&quot-u&quot, dest=&quotmut_up&quot,
      default=0, type=&quotint&quot,
      help=&quotNucleotides upstream of center sequence to mutate [Default: %default]&quot)
  <a id="change">(options, args) = parser.parse_args()</a>

  if len(args) == 2:
    &#47&#47 single worker
    <a id="change">model_file = args[0]</a>
    <a id="change">bed_file = args[1]</a>

  elif len(args) == 3:
    &#47&#47 master script
    <a id="change">options_pkl_file = args[0]</a>
    <a id="change">model_file = args[1]</a>
    <a id="change">bed_file = args[2]</a>

    &#47&#47 load options
    <a id="change">options_pkl = open(options_pkl_file, &quotrb&quot)</a>
    <a id="change">options = pickle.load(options_pkl)</a>
    <a id="change">options_pkl</a><a id="change">.close()</a>

  elif len(args) == 4:
    &#47&#47 multi worker
    <a id="change">options_pkl_file = args[0]</a>
    <a id="change">model_file = args[1]</a>
    <a id="change">bed_file = args[2]</a>
    <a id="change">worker_index = int(args[3])</a>

    &#47&#47 load options
    <a id="change">options_pkl = open(options_pkl_file, &quotrb&quot)</a>
    <a id="change">options = pickle.load(options_pkl)</a>
    <a id="change">options_pkl</a><a id="change">.close()</a>

    &#47&#47 update output directory
    <a id="change">options.out_dir = &quot%s/job%d&quot % (options.out_dir, worker_index)</a>

  else:
    parser.error(&quotMust provide parameter and model files and BED file&quot)

  if not os.path.isdir(options.out_dir):
    os.mkdir(options.out_dir)

  if options.mut_up &gt; 0 or options.mut_down &gt; 0:
    <a id="change">options.mut_len = options.mut_up + options.mut_down</a>
  else:
    assert(options.mut_len &gt; 0)
    <a id="change">options.mut_up = options.mut_len // 2</a>
    <a id="change">options.mut_down = options.mut_len - options.mut_up</a>

  <a id="change">options.shifts = [int(shift) for shift in options.shifts.split(&quot,&quot)]</a>

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 setup model

  <a id="change">seqnn_model = tf.saved_model.load(model_file).model</a>

  &#47&#47 query num model targets 
  <a id="change">seq_length = seqnn_model.predict_on_batch.input_signature[0].shape[1]</a>
  <a id="change">null_1hot = np.zeros((1,seq_length,4))</a>
  <a id="change">null_preds = seqnn_model.predict_on_batch(null_1hot)</a>
  <a id="change">null_preds = null_preds[options.species].numpy()</a>
  <a id="change">_, preds_length, num_targets = null_preds.shape</a>

  &#47&#47 read targets
  if options.targets_file is None:
    <a id="change">targets_mask = np.ones(num_targets, dtype=&quotfloat32&quot)</a>
  else:
    <a id="change">targets_df = pd.read_table(options.targets_file, index_col=0)</a>
    <a id="change">targets_mask = np.zeros(num_targets, dtype=&quotfloat32&quot)</a>
    <a id="change">targets_mask[targets_df.index] = 1.0</a>

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 sequence dataset

  &#47&#47 read sequences from BED
  <a id="change">seqs_dna, seqs_coords = bed.make_bed_seqs(
    bed_file, options.genome_fasta, seq_length, stranded=True)</a>

  &#47&#47 filter for worker SNPs
  if options.processes is not None:
    <a id="change">worker_bounds = np.linspace(0, len(seqs_dna), options.processes+1, dtype=&quotint&quot)</a>
    <a id="change">seqs_dna = seqs_dna[worker_bounds[worker_index]:worker_bounds[worker_index+1]]</a>
    <a id="change">seqs_coords = seqs_coords[worker_bounds[worker_index]:worker_bounds[worker_index+1]]</a>

  <a id="change">num_seqs = len(seqs_dna)</a>

  &#47&#47 determine mutation region limits
  <a id="change">seq_mid = seq_length // 2</a>
  <a id="change">mut_start = seq_mid - options.mut_up</a>
  <a id="change">mut_end = mut_start + options.mut_len</a>

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 setup output

  <a id="change">scores_h5_file = &quot%s/scores.h5&quot % options.out_dir</a>
  if os.path.isfile(scores_h5_file):
    os.remove(scores_h5_file)
  <a id="change">scores_h5 = h5py.File(scores_h5_file, &quotw&quot)</a>
  scores_h5.create_dataset(&quotseqs&quot, dtype=&quotbool&quot,
      shape=(num_seqs, options.mut_len, 4))
  scores_h5.create_dataset(&quotgrads&quot, dtype=&quotfloat16&quot,
      shape=(num_seqs, options.mut_len, 4))

  &#47&#47 store mutagenesis sequence coordinates
  <a id="change">scores_chr = []</a>
  <a id="change">scores_start = []</a>
  <a id="change">scores_end = []</a>
  <a id="change">scores_strand = []</a>
  for seq_chr, seq_start, seq_end, seq_strand in seqs_coords:
    scores_chr.append(seq_chr)
    scores_strand.append(seq_strand)
    if seq_strand == &quot+&quot:
      <a id="change">score_start = seq_start + mut_start</a>
      <a id="change">score_end = score_start + options.mut_len</a>
    else:
      <a id="change">score_end = seq_end - mut_start</a>
      <a id="change">score_start = score_end - options.mut_len</a>
    scores_start.append(score_start)
    scores_end.append(score_end)

  scores_h5.create_dataset(&quotchr&quot, data=np.array(scores_chr, dtype=&quotS&quot))
  scores_h5.create_dataset(&quotstart&quot, data=np.array(scores_start))
  scores_h5.create_dataset(&quotend&quot, data=np.array(scores_end))
  scores_h5.create_dataset(&quotstrand&quot, data=np.array(scores_strand, dtype=&quotS&quot))

  <a id="change">preds_per_seq = 1 + 3*options.mut_len</a>

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 predict scores, write output

  &#47&#47 find center
  if options.slice_center:
    <a id="change">center_start = preds_length // 2</a>
    if preds_length % 2 == 0:
      <a id="change">center_end = center_start + 2</a>
    else:
      <a id="change">center_end = center_start + 1</a>
  else:
    <a id="change">center_start = 0</a>
    <a id="change">center_end = preds_length</a>    

  <a id="change">si = 0</a>
  for seq_dna in seqs_dna:
    print(&quotPredicting %d&quot % si, flush=True)
    <a id="change">seq_1hot = dna_io.dna_1hot(seq_dna)</a>

    &#47&#47 &#47&#47 forward
    &#47&#47 seq_1hot_tf = tf.convert_to_tensor(seq_1hot, dtype=tf.float32)[tf.newaxis]

    &#47&#47 &#47&#47 compute gradients
    &#47&#47 grad_f = input_gradients(seqnn_model, seq_1hot_tf, targets_mask,
    &#47&#47   center_start, center_end, options.species).numpy()

    &#47&#47 if options.rc:
    &#47&#47   &#47&#47 reverse
    &#47&#47   seq_1hot_r = dna_io.hot1_rc(seq_1hot)
    &#47&#47   seq_1hot_tf = tf.convert_to_tensor(seq_1hot_r, dtype=tf.float32)[tf.newaxis]

    &#47&#47   &#47&#47 compute gradients
    &#47&#47   grad_r = input_gradients(seqnn_model, seq_1hot_tf, targets_mask,
    &#47&#47     center_start, center_end, options.species).numpy()
    &#47&#47   grad_r = dna_io.hot1_rc(grad_r)

    &#47&#47   &#47&#47 average
    &#47&#47   grad = (grad_f + grad_r) / 2

    &#47&#47 else:
    &#47&#47   grad = grad_f

    <a id="change">grad_ens = []</a>
    for shift in options.shifts:
      <a id="change">seq_1hot_aug = dna_io.hot1_augment(seq_1hot, shift=shift)</a>
      <a id="change">seq_1hot_tf = tf.convert_to_tensor(seq_1hot_aug, dtype=tf.float32)[tf.newaxis]</a>
      <a id="change">grad_aug = input_gradients(seqnn_model, seq_1hot_tf, targets_mask,
          center_start, center_end, options.species).numpy()</a>
      <a id="change">grad_aug = dna_io.hot1_augment(grad_aug, fwdrc=True, shift=-shift)</a>
      grad_ens.append(grad_aug)
      
      if options.rc:
        <a id="change">seq_1hot_aug = dna_io.hot1_rc(seq_1hot_aug)</a>
        <a id="change">seq_1hot_tf = tf.convert_to_tensor(seq_1hot_aug, dtype=tf.float32)[tf.newaxis]</a>
        <a id="change">grad_aug = input_gradients(seqnn_model, seq_1hot_tf, targets_mask,
          center_start, center_end, options.species).numpy()</a>
        <a id="change">grad_aug = dna_io.hot1_augment(grad_aug, fwdrc=False, shift=-shift)</a>
        grad_ens.append(grad_aug)

    <a id="change">grad = np.array(grad_ens).mean(axis=0)</a>

    &#47&#47 write to HDF5
    <a id="change">scores_h5[&quotseqs&quot][si] = seq_1hot[mut_start:mut_end]</a>
    <a id="change">scores_h5[&quotgrads&quot][si] = grad[mut_start:mut_end].astype(&quotfloat16&quot)</a>

    si += 1

  &#47&#47 close output HDF5
  scores_h5.close()
    

@tf.function
def input_gradients(seqnn_model, seq_1hot_tf, targets_mask, pos_start, pos_end, species):
  &#47&#47 watch prediction
  targets_mass = tf.reduce_sum(targets_mask)
  with tf.GradientTape() as tape:
    tape.watch(seq_1hot_tf)
    pred_raw = seqnn_model.predict_on_batch(seq_1hot_tf)[species]

    &#47&#47 mean across targets
    raw2_pred = tf.reduce_sum(targets_mask[tf.newaxis]*pred_raw / targets_mass, axis=-1)

    &#47&#47 slice center and take mean
    grad_pred = tf.reduce_sum(raw2_pred[:,pos_start:pos_end], axis=-1)

  &#47&#47 compute gradient
  grad = tape.gradient(grad_pred, seq_1hot_tf)
  grad = tf.squeeze(grad, axis=0)

  &#47&#47 zero mean each position
  grad = grad - tf.reduce_mean(grad, axis=-1, keepdims=True)

  return grad

&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
&#47&#47 __main__
&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
if __name__ == &quot__main__&quot:
  main()
</code></pre>