<link rel="stylesheet" href="../../..//default.css">
<script src="../../..//highlight.pack.js"></script> 
<script>hljs.initHighlightingOnLoad();</script>
<html><pre><code class='java'>
<a href="https://github.com/calico/basenji/blob/master/bin/sonnet_sat_bed.py#L54">GitHubLink</a>


<a href="https://github.com/maldil/basenji/blob/master/bin/sonnet_sat_bed.py#L54">GitMyHubLink</a>

&#47&#47!/usr/bin/env python
&#47&#47 Copyright 2017 Calico LLC
&#47&#47
&#47&#47 Licensed under the Apache License, Version 2.0 (the "License");
&#47&#47 you may not use this file except in compliance with the License.
&#47&#47 You may obtain a copy of the License at
&#47&#47
&#47&#47     https://www.apache.org/licenses/LICENSE-2.0
&#47&#47
&#47&#47 Unless required by applicable law or agreed to in writing, software
&#47&#47 distributed under the License is distributed on an "AS IS" BASIS,
&#47&#47 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
&#47&#47 See the License for the specific language governing permissions and
&#47&#47 limitations under the License.
&#47&#47 =========================================================================
from __future__ import print_function

from optparse import OptionParser

import gc
import json
import os
import pdb
import pickle
from queue import Queue
import random
import sys
from threading import Thread

import h5py
import numpy as np
import pandas as pd
import pysam
import tensorflow as tf

if tf.__version__[0] == &quot1&quot:
  tf.compat.v1.enable_eager_execution()

from basenji import bed
from basenji import dna_io
from basenji import seqnn
from basenji import stream
from basenji_sat_bed import satmut_gen, ScoreWorker

&quot&quot&quot
sonnet_sat_bed.py

Perform an in silico saturation mutagenesis of sequences in a BED file.
&quot&quot&quot

&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
&#47&#47 main
&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
def main():
  <a id="change">usage = &quotusage: %prog [options] &lt;model&gt; &lt;bed_file&gt;&quot</a>
  <a id="change">parser = OptionParser(usage)</a>
  parser.add_option(&quot-a&quot, dest=&quotreturn_augmentations&quot,
      default=False, action=&quotstore_true&quot,
      help=&quotReturn all augmentations, rather than average [Default: %default]&quot)
  parser.add_option(&quot-b&quot, dest=&quotbatch_size&quot,
      default=4, type=&quotint&quot,
      help=&quotBatch size [Default: %default]&quot)
  parser.add_option(&quot-d&quot, dest=&quotmut_down&quot,
      default=0, type=&quotint&quot,
      help=&quotNucleotides downstream of center sequence to mutate [Default: %default]&quot)
  parser.add_option(&quot-f&quot, dest=&quotgenome_fasta&quot,
      default=None,
      help=&quotGenome FASTA for sequences [Default: %default]&quot)
  parser.add_option(&quot-l&quot, dest=&quotmut_len&quot,
      default=0, type=&quotint&quot,
      help=&quotLength of center sequence to mutate [Default: %default]&quot)
  parser.add_option(&quot-o&quot, dest=&quotout_dir&quot,
      default=&quotsat_mut&quot, help=&quotOutput directory [Default: %default]&quot)
  parser.add_option(&quot--plots&quot, dest=&quotplots&quot,
      default=False, action=&quotstore_true&quot,
      help=&quotMake heatmap plots [Default: %default]&quot)
  parser.add_option(&quot-p&quot, dest=&quotprocesses&quot,
      default=None, type=&quotint&quot,
      help=&quotNumber of processes, passed by multi script&quot)
  parser.add_option(&quot--rc&quot, dest=&quotrc&quot,
      default=False, action=&quotstore_true&quot,
      help=&quotEnsemble forward and reverse complement predictions [Default: %default]&quot)
  parser.add_option(&quot--shifts&quot, dest=&quotshifts&quot,
      default=&quot0&quot,
      help=&quotEnsemble prediction shifts [Default: %default]&quot)
  parser.add_option(&quot--species&quot, dest=&quotspecies&quot,
      default=&quothuman&quot)
  parser.add_option(&quot--stats&quot, dest=&quotsad_stats&quot,
      default=&quotsum&quot,
      help=&quotComma-separated list of stats to save (sum/center/scd). [Default: %default]&quot)
  parser.add_option(&quot-t&quot, dest=&quottargets_file&quot,
      default=None, type=&quotstr&quot,
      help=&quotFile specifying target indexes and labels in table format&quot)
  parser.add_option(&quot-u&quot, dest=&quotmut_up&quot,
      default=0, type=&quotint&quot,
      help=&quotNucleotides upstream of center sequence to mutate [Default: %default]&quot)
  <a id="change">(options, args) = parser.parse_args()</a>

  if len(args) == 2:
    &#47&#47 single worker
    <a id="change">model_file = args[0]</a>
    <a id="change">bed_file = args[1]</a>

  elif len(args) == 3:
    &#47&#47 master script
    <a id="change">options_pkl_file = args[0]</a>
    <a id="change">model_file = args[1]</a>
    <a id="change">bed_file = args[2]</a>

    &#47&#47 load options
    <a id="change">options_pkl = open(options_pkl_file, &quotrb&quot)</a>
    <a id="change">options = pickle.load(options_pkl)</a>
    <a id="change">options_pkl</a><a id="change">.close()</a>

  elif len(args) == 4:
    &#47&#47 multi worker
    <a id="change">options_pkl_file = args[0]</a>
    <a id="change">model_file = args[1]</a>
    <a id="change">bed_file = args[2]</a>
    <a id="change">worker_index = int(args[3])</a>

    &#47&#47 load options
    <a id="change">options_pkl = open(options_pkl_file, &quotrb&quot)</a>
    <a id="change">options = pickle.load(options_pkl)</a>
    <a id="change">options_pkl</a><a id="change">.close()</a>

    &#47&#47 update output directory
    <a id="change">options.out_dir = &quot%s/job%d&quot % (options.out_dir, worker_index)</a>

  else:
    parser.error(&quotMust provide parameter and model files and BED file&quot)

  if not os.path.isdir(options.out_dir):
    os.mkdir(options.out_dir)

  <a id="change">options.shifts = [int(shift) for shift in options.shifts.split(&quot,&quot)]</a>
  <a id="change">num_augmentations = len(options.shifts) + int(options.rc)*len(options.shifts)</a>
  <a id="change">options.sad_stats = [sad_stat.lower() for sad_stat in options.sad_stats.split(&quot,&quot)]</a>

  if options.mut_up &gt; 0 or options.mut_down &gt; 0:
    <a id="change">options.mut_len = options.mut_up + options.mut_down</a>
  else:
    assert(options.mut_len &gt; 0)
    <a id="change">options.mut_up = options.mut_len // 2</a>
    <a id="change">options.mut_down = options.mut_len - options.mut_up</a>

  &#47&#47 read targets
  if options.targets_file is None:
    <a id="change">target_slice = None</a>
  else:
    <a id="change">targets_df = pd.read_table(options.targets_file, index_col=0)</a>
    <a id="change">target_slice = targets_df.index</a>

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 setup model

  <a id="change">seqnn_model = tf.saved_model.load(model_file).model</a>

  &#47&#47 query num model targets 
  <a id="change">seq_length = seqnn_model.predict_on_batch.input_signature[0].shape[1]</a>
  <a id="change">null_1hot = np.zeros((1,seq_length,4))</a>
  <a id="change">null_preds = seqnn_model.predict_on_batch(null_1hot)</a>
  <a id="change">null_preds = null_preds[options.species].numpy()</a>
  <a id="change">_, preds_length, num_targets = null_preds.shape</a>

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 sequence dataset

  &#47&#47 read sequences from BED
  <a id="change">seqs_dna, seqs_coords = bed.make_bed_seqs(
    bed_file, options.genome_fasta, seq_length, stranded=True)</a>

  &#47&#47 filter for worker SNPs
  if options.processes is not None:
    <a id="change">worker_bounds = np.linspace(0, len(seqs_dna), options.processes+1, dtype=&quotint&quot)</a>
    <a id="change">seqs_dna = seqs_dna[worker_bounds[worker_index]:worker_bounds[worker_index+1]]</a>
    <a id="change">seqs_coords = seqs_coords[worker_bounds[worker_index]:worker_bounds[worker_index+1]]</a>

  <a id="change">num_seqs = len(seqs_dna)</a>

  &#47&#47 determine mutation region limits
  <a id="change">seq_mid = seq_length // 2</a>
  <a id="change">mut_start = seq_mid - options.mut_up</a>
  <a id="change">mut_end = mut_start + options.mut_len</a>

  &#47&#47 make sequence generator
  <a id="change">seqs_gen = satmut_gen(seqs_dna, mut_start, mut_end)</a>

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 setup output

  <a id="change">scores_h5_file = &quot%s/scores.h5&quot % options.out_dir</a>
  if os.path.isfile(scores_h5_file):
    os.remove(scores_h5_file)
  <a id="change">scores_h5 = h5py.File(scores_h5_file, &quotw&quot)</a>
  scores_h5.create_dataset(&quotseqs&quot, dtype=&quotbool&quot,
      shape=(num_seqs, options.mut_len, 4))
  for sad_stat in options.sad_stats:
    if options.return_augmentations:
      scores_h5.create_dataset(sad_stat, dtype=&quotfloat16&quot,
          shape=(num_seqs, options.mut_len, 4, num_targets, num_augmentations))
    else:
      scores_h5.create_dataset(sad_stat, dtype=&quotfloat16&quot,
          shape=(num_seqs, options.mut_len, 4, num_targets))

  &#47&#47 store mutagenesis sequence coordinates
  <a id="change">scores_chr = []</a>
  <a id="change">scores_start = []</a>
  <a id="change">scores_end = []</a>
  <a id="change">scores_strand = []</a>
  for seq_chr, seq_start, seq_end, seq_strand in seqs_coords:
    scores_chr.append(seq_chr)
    scores_strand.append(seq_strand)
    if seq_strand == &quot+&quot:
      <a id="change">score_start = seq_start + mut_start</a>
      <a id="change">score_end = score_start + options.mut_len</a>
    else:
      <a id="change">score_end = seq_end - mut_start</a>
      <a id="change">score_start = score_end - options.mut_len</a>
    scores_start.append(score_start)
    scores_end.append(score_end)

  scores_h5.create_dataset(&quotchr&quot, data=np.array(scores_chr, dtype=&quotS&quot))
  scores_h5.create_dataset(&quotstart&quot, data=np.array(scores_start))
  scores_h5.create_dataset(&quotend&quot, data=np.array(scores_end))
  scores_h5.create_dataset(&quotstrand&quot, data=np.array(scores_strand, dtype=&quotS&quot))

  <a id="change">preds_per_seq = 1 + 3*options.mut_len</a>

  <a id="change">score_threads = []</a>
  <a id="change">score_queue = Queue()</a>
  for i in range(1):
    <a id="change">sw = ScoreWorker(score_queue, scores_h5, options.sad_stats,
                     mut_start, mut_end, return_augm=options.return_augmentations)</a>
    sw.start()
    score_threads.append(sw)

  &#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
  &#47&#47 predict scores, write output

  &#47&#47 find center
  <a id="change">center_start = preds_length // 2</a>
  if preds_length % 2 == 0:
    <a id="change">center_end = center_start + 2</a>
  else:
    <a id="change">center_end = center_start + 1</a>

  &#47&#47 initialize predictions stream
  <a id="change">preds_stream = stream.PredStreamSonnet(seqnn_model, seqs_gen,
    rc=options.rc, shifts=options.shifts, species=options.species,
    batch_size=options.batch_size, return_augm=options.return_augmentations)</a>

  &#47&#47 predictions index
  <a id="change">pi = 0</a>

  for si in range(num_seqs):
    print(&quotPredicting %d&quot % si, flush=True)

    &#47&#47 collect sequence predictions
    <a id="change">seq_preds_sum = []</a>
    <a id="change">seq_preds_center = []</a>
    <a id="change">seq_preds_scd = []</a>
    <a id="change">preds_mut0 = preds_stream[pi]</a>
    for spi in range(preds_per_seq):
      <a id="change">preds_mut = preds_stream[pi]</a>
      <a id="change">preds_sum = preds_mut.sum(axis=0)</a>
      seq_preds_sum.append(preds_sum)
      if &quotcenter&quot in options.sad_stats:
        <a id="change">preds_center = preds_mut[center_start:center_end,:].sum(axis=0)</a>
        seq_preds_center.append(preds_center)
      if &quotscd&quot in options.sad_stats:
        <a id="change">preds_scd = np.sqrt(((preds_mut-preds_mut0)**2).sum(axis=0))</a>
        seq_preds_scd.append(preds_scd)
      pi += 1
    <a id="change">seq_preds_sum = np.array(seq_preds_sum)</a>
    <a id="change">seq_preds_center = np.array(seq_preds_center)</a>
    <a id="change">seq_preds_scd = np.array(seq_preds_scd)</a>

    &#47&#47 wait for previous to finish
    score_queue.join()

    &#47&#47 queue sequence for scoring
    seq_pred_stats = (seq_preds_sum, seq_preds_center, seq_preds_scd)
    score_queue.put((seqs_dna[si], seq_pred_stats, si))
    
    &#47&#47 queue sequence for plotting
    if options.plots:
      plot_queue.put((seqs_dna[si], seq_preds_sum, si))

    gc.collect()

  &#47&#47 finish queue
  print(&quotWaiting for threads to finish.&quot, flush=True)
  score_queue.join()

  &#47&#47 close output HDF5
  scores_h5.close()


&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
&#47&#47 __main__
&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47&#47
if __name__ == &quot__main__&quot:
  main()
</code></pre>