<link rel="stylesheet" href="../../../../../..//default.css">
<script src="../../../../../..//highlight.pack.js"></script> 
<script>hljs.initHighlightingOnLoad();</script>
<html><pre><code class='java'>
<a href="https://github.com/10XGenomics/cellranger/blob/master/tenkit/lib/python/tenkit/align.py#L51">GitHubLink</a>


<a href="https://github.com/maldil/cellranger/blob/master/tenkit/lib/python/tenkit/align.py#L51">GitMyHubLink</a>

&#47&#47!/usr/bin/env python
&#47&#47
&#47&#47 Copyright (c) 2014 10X Genomics, Inc. All rights reserved.
&#47&#47
&#47&#47 Run BWA and sort
&#47&#47

import os
import tenkit.bam as tk_bam
import tenkit.fasta as tk_fasta
from exceptions import NotSupportedException
import subprocess
import log_subprocess

class Aligner(object):
    
    Class for taking in a fastq file and running an aligner on it.
    
    def __init__(self, in_fastq_file_name, output_name):
        self.in_fastq_file_name = in_fastq_file_name
        self.output_name = output_name

    def output_alignment(self, aligner=None, aligner_params=None, paired_end=True, genome=None, num_threads=16, read_group_header=None):
         Runs the given aligner on the input bam file
        If aligner is None, just copies the bam file to the new directory and then sorts and indexes it.
        
        if aligner:
            if aligner == &quotbwa&quot:
                ref_fasta = aligner_params[&quotref_fasta&quot]
                algorithm = aligner_params[&quotalgorithm&quot]
                if paired_end:
                    bwa_align_paired(ref_fasta, self.in_fastq_file_name, self.output_name, algorithm=algorithm, num_threads=num_threads, read_group_header=read_group_header)
                else:
                    bwa_align_unpaired(ref_fasta, self.in_fastq_file_name, self.output_name, algorithm=algorithm, num_threads=num_threads, read_group_header=read_group_header)
            else:
                raise NotSupportedException(&quotAlignment method not supported: &quot + aligner)
        else:
            raise NotSupportedException(&quotMust specify an aligner&quot)

def bwa_read_group_arg(read_group, sample):
    &quot&quot&quot The -R argument to BWA must be escaped properly so that argument string contains two character \\t string, rather than actual tabs &quot&quot&quot
    return &quot@RG\\tID:&quot + read_group + &quot\\tSM:&quot + sample

def bwa_index_ref(ref_fasta):
     Creates index of reference for bwa.  ref_fasta should be path to the reference fasta
    Only needs to be called once per reference.  Creates index files in the same directory as the
    reference
    
    log_subprocess.check_call([&quotbwa&quot, &quotindex&quot, &quot-a&quot, &quotbwtsw&quot, ref_fasta])

def bwa_align_unpaired(ref_fasta, read_fastq, out_name, algorithm=&quotALN&quot, max_hits=None, read_group_header=None, num_threads=24):
     Runs bwa aligner on reads without using paired-information (using bam as input format).
    
    assert(type(read_fastq) != list)

    if read_group_header is None:
        read_group_header = tk_bam.make_rg_header()

    if algorithm == &quotMEM&quot:
        &#47&#47 Temp file names
        sam_name = out_name + &quot.sam&quot

        <a id="change">sam_out_file = open(sam_name, &quotw&quot)</a>
        log_subprocess.check_call([&quotbwa&quot, &quotmem&quot, &quot-t&quot, str(num_threads), &quot-M&quot, &quot-R&quot, read_group_header, ref_fasta, read_fastq], stdout=sam_out_file)
        <a id="change">sam_out_file</a><a id="change">.close()</a>

        &#47&#47 Create final bam file from the sam file
        tk_bam.convert_to_bam(sam_name, out_name)

        &#47&#47 Remove temp files
        subprocess.check_call([&quotrm&quot, sam_name])

    elif algorithm == &quotALN&quot:
        &#47&#47 Temp file names
        sam_name = out_name + &quot.sam&quot
        index_name = out_name + &quot.sai&quot

        <a id="change">sam_out_file = open(sam_name, &quotw&quot)</a>
        <a id="change">index_file = open(index_name, &quotw&quot)</a>
        log_subprocess.check_call([&quotbwa&quot, &quotaln&quot, &quot-t&quot, str(num_threads), ref_fasta, read_fastq], stdout=index_file)
        <a id="change">index_file</a><a id="change">.close()</a>
        if max_hits:
            log_subprocess.check_call([&quotbwa&quot, &quotsamse&quot, &quot-n&quot, str(max_hits), ref_fasta, index_name, read_fastq], stdout=sam_out_file)
        else:
            log_subprocess.check_call([&quotbwa&quot, &quotsamse&quot, ref_fasta, index_name, read_fastq], stdout=sam_out_file)
        <a id="change">sam_out_file</a><a id="change">.close()</a>

        &#47&#47 Create final bam file from the sam file
        tk_bam.convert_to_bam(sam_name, out_name)

        &#47&#47 Remove temp files
        subprocess.check_call([&quotrm&quot, index_name])
        subprocess.check_call([&quotrm&quot, sam_name])
    else:
        raise NotSupportedException(&quotUnsupported bwa algorithm: &quot + algorithm)


def bwa_align_paired(ref_fasta, read_fastq, out_name, algorithm=&quotALN&quot, max_hits=None, read_group_header=None, num_threads=24):
    Runs bwa paired-end aligner on reads using paired-end information
    Algorithm choices are currently
    MEM: Maximal Exact Matching (better for longer reads)
    ALN: Better for longer reads
    Haven&quott yet implemented BWA-SW
    Currently assumes the input read_fastq is in interleaved format, i.e. the reads of a pair
    are alternating.
    
    if read_group_header is None:
        read_group_header = tk_bam.make_rg_header()

    if algorithm == &quotMEM&quot:
        devnull = open(os.devnull, &quotw&quot)
        if type(read_fastq) == list:
            assert(len(read_fastq) == 2)
            &#47&#47&#47&#47 This restricts to primary alignments only
            out_file = open(out_name, &quotw&quot)
            ps = log_subprocess.Popen([&quotbwa&quot, &quotmem&quot, &quot-t&quot, str(num_threads), &quot-M&quot,  &quot-R&quot, read_group_header, ref_fasta, read_fastq[0], read_fastq[1]],
                                      stdout=subprocess.PIPE, stderr=devnull)
            &#47&#47log_subprocess.check_call([&quotsamtools&quot, &quotview&quot, &quot-bSh&quot, &quot-&quot], stdin=ps.stdout, stdout=out_file) &#47&#47 restore once bug fixed
            errors_file = open(out_name + &quot_ERRORS&quot, &quotw&quot)
            log_subprocess.check_call([&quotsamtools&quot, &quotview&quot, &quot-bSh&quot, &quot-&quot], stdin=ps.stdout, stdout=out_file, stderr=errors_file)
            out_file.close()
            errors_file.close()
        else:
            &#47&#47&#47&#47 This restricts to primary alignments only
            out_file = open(out_name, &quotw&quot)
            ps = log_subprocess.Popen([&quotbwa&quot, &quotmem&quot, &quot-p&quot, &quot-t&quot, str(num_threads), &quot-M&quot,  &quot-R&quot, read_group_header, ref_fasta, read_fastq],
                                      stdout=subprocess.PIPE, stderr=devnull)
            &#47&#47log_subprocess.check_call([&quotsamtools&quot, &quotview&quot, &quot-bSh&quot, &quot-&quot], stdin=ps.stdout, stdout=out_file) &#47&#47 restore once bug fixed
            errors_file = open(out_name + &quot_ERRORS&quot, &quotw&quot)
            log_subprocess.check_call([&quotsamtools&quot, &quotview&quot, &quot-bSh&quot, &quot-&quot], stdin=ps.stdout, stdout=out_file, stderr=errors_file)
            out_file.close()
            errors_file.close()
        devnull.close()

    elif algorithm == &quotALN&quot:
        &#47&#47 Temp file names
        temp_fastq_name1 = out_name + &quot1.fastq&quot
        temp_fastq_name2 = out_name + &quot2.fastq&quot
        index_name_1 = out_name + &quot1.sai&quot
        index_name_2 = out_name + &quot2.sai&quot
        sam_name = out_name + &quot.sam&quot

        &#47&#47 Create the temp non-interleaved files
        in_fastq = open(read_fastq, &quotr&quot)
        temp_fastq1 = open(temp_fastq_name1, &quotw&quot)
        temp_fastq2 = open(temp_fastq_name2, &quotw&quot)
        tk_fasta.uninterleave_fastq(in_fastq, temp_fastq1, temp_fastq2)
        temp_fastq1.close()
        temp_fastq2.close()

        &#47&#47 Create the bwa index files
        index_file_1 = open(index_name_1, &quotw&quot)
        index_file_2 = open(index_name_2, &quotw&quot)
        log_subprocess.check_call([&quotbwa&quot, &quotaln&quot, &quot-t&quot, str(num_threads), ref_fasta, temp_fastq_name1], stdout=index_file_1)
        log_subprocess.check_call([&quotbwa&quot, &quotaln&quot, &quot-t&quot, str(num_threads), ref_fasta, temp_fastq_name2], stdout=index_file_2)
        index_file_1.close()
        index_file_2.close()

        &#47&#47 Create the sorted SAM file
        sam_out_file = open(sam_name, &quotw&quot)
        if max_hits:
            log_subprocess.check_call([&quotbwa&quot, &quotsampe&quot, &quot-n&quot, str(max_hits), ref_fasta, index_name_1, index_name_2, temp_fastq_name1, temp_fastq_name2], stdout=sam_out_file)
        else:
            log_subprocess.check_call([&quotbwa&quot, &quotsampe&quot, ref_fasta, index_name_1, index_name_2, temp_fastq_name1, temp_fastq_name2], stdout=sam_out_file)

        sam_out_file.close()

        &#47&#47 Create final bam file from the sam file
        tk_bam.convert_to_bam(sam_name, out_name)

        &#47&#47 Clean up temporary files
        subprocess.check_call([&quotrm&quot, temp_fastq_name1])
        subprocess.check_call([&quotrm&quot, temp_fastq_name2])
        subprocess.check_call([&quotrm&quot, index_name_1])
        subprocess.check_call([&quotrm&quot, index_name_2])
        subprocess.check_call([&quotrm&quot, sam_name])
    else:
        raise NotSupportedException(&quotUnsupported bwa algorithm: &quot + algorithm)
</code></pre>